src/env.rs

```diff
@@ -0,0 +1,278 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An environment which is passed downward into the widget tree.
+
+// TODO: remove this
+#![allow(unused)]
+
+use std::collections::HashMap;
+use std::ops::{Deref, Index};
+
+use crate::piet::Color;
+
+use crate::Data;
+
+// Discussion question: since druid is pretty rigorously single-threaded,
+// maybe we should use Rc instead of Arc.
+use std::sync::Arc;
+
+/// An environment passed down through all widget traversals.
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone)]
+pub struct Env(Arc<EnvImpl>);
+
+struct EnvImpl {
+    theme: EnvValue,
+    // This is always a map, maybe should be a map here.
+    values: EnvValue,
+}
+
+/// A value that can be stored in an environment.
+///
+/// This is an enum with a variety of common types.
+#[derive(Clone)]
+pub enum EnvValue {
+    Float(f64),
+    Bool(bool),
+    Color(Color),
+    String(String),
+    List(Arc<Vec<EnvValue>>),
+    Map(Arc<HashMap<String, EnvValue>>),
+    Null,
+}
+
+impl Env {
+    /// Create an environment with a theme but an empty values map.
+    pub fn from_theme(theme: EnvValue) -> Env {
+        Env(Arc::new(EnvImpl {
+            theme,
+            values: EnvValue::empty_map(),
+        }))
+    }
+
+    /// Get a theme value.
+    ///
+    /// The value will be EnvValue::Null if the key is not present.
+    // TODO: make key more flexible
+    pub fn theme_value(&self, key: &str) -> &EnvValue {
+        &self.0.theme[key]
+    }
+
+    /// Get a theme color.
+    ///
+    /// Prints a debug warning and returns a garish color if the key
+    /// isn't present or is of the wrong type.
+    pub fn theme_color(&self, key: &str) -> Color {
```

Comment:
I don't love these two methods either, but it's kind of a finicky problem.

I think getting this right demands we think more about what a theme is. There's a very reasonable case to be made that a theme is either a fixed set of keys/values (as in: it is an invariant that a given set of keys exists, and each key has a known value type), *or* a theme is that *plus* some custom keys and values provided by the user.

In either case, (for themes, maybe not for `env` more generally) keys and the type of their values should always be known at compile time.

In my environment branch, I was playing around with various ways of encoding this in the type system. The best approach I've found is to encode the type of the value in the key itself, like:

```rust
pub struct Key<T> {
    key: &'static str,
    value_type: PhantomData<T>,
}

pub const TEXT: Key<Color> = Key::new("io.xi-editor.text_color");

let color: Color = theme.get(&TEXT); // Color or panic
```

For all built-in keys, we can guarantee that they are of the correct type. This also lets us do things like disallow overwriting a key with a value of a different type.

Another option is to have a fn `get`, that is generic over it's return type:

```rust
fn get<T: FromEnvValue>(key: &Key) -> T;
```

where `FromEnvValue` is a trait implemented by the raw `EnvValue` types.